pkg/controller/jobs/pod/pod_multikueue_adapter.go

```diff
@@ -0,0 +1,134 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pod
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/constants"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	"sigs.k8s.io/kueue/pkg/features"
+	"sigs.k8s.io/kueue/pkg/util/api"
+	clientutil "sigs.k8s.io/kueue/pkg/util/client"
+)
+
+type multikueueAdapter struct{}
+
+var _ jobframework.MultiKueueAdapter = (*multikueueAdapter)(nil)
+
+func (b *multikueueAdapter) SyncJob(ctx context.Context, localClient client.Client, remoteClient client.Client, key types.NamespacedName, workloadName, origin string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	localPod := corev1.Pod{}
+	err := localClient.Get(ctx, key, &localPod)
+	if err != nil {
+		return err
+	}
+
+	remotePod := corev1.Pod{}
+	err = remoteClient.Get(ctx, key, &remotePod)
+	if client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// the remote pod exists
+	if err == nil {
+		if FromObject(&localPod).IsSuspended() {
+			// Ensure the pod is unsuspended before updating its status; otherwise, it will fail when patching the spec.
+			log.V(2).Info("Skipping the sync since the local pod is still suspended")
+			return nil
+		}
+
+		remoteFinished := false
+		if remotePod.Status.Phase == corev1.PodSucceeded || remotePod.Status.Phase == corev1.PodFailed {
+			remoteFinished = true
+		}
+		if remoteFinished {
+			return clientutil.PatchStatus(ctx, localClient, &localPod, func() (bool, error) {
+				localPod.Status = remotePod.Status
+				return true, nil
+			})
+		}
+		return nil
+	}
+
+	remotePod = corev1.Pod{
+		ObjectMeta: api.CloneObjectMetaForCreation(&localPod.ObjectMeta),
+		Spec:       *localPod.Spec.DeepCopy(),
+	}
+
+	// add the prebuilt workload
+	if remotePod.Labels == nil {
+		remotePod.Labels = map[string]string{}
+	}
+	remotePod.Labels[constants.PrebuiltWorkloadLabel] = workloadName
+	remotePod.Labels[kueue.MultiKueueOriginLabel] = origin
+
+	return remoteClient.Create(ctx, &remotePod)
+}
+
+func (b *multikueueAdapter) DeleteRemoteObject(ctx context.Context, remoteClient client.Client, key types.NamespacedName) error {
+	pod := corev1.Pod{}
+	err := remoteClient.Get(ctx, key, &pod)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	return client.IgnoreNotFound(remoteClient.Delete(ctx, &pod))
+}
+
+func (b *multikueueAdapter) KeepAdmissionCheckPending() bool {
+	return !features.Enabled(features.MultiKueueBatchJobWithManagedBy)
```

Comment:
This feature gate probably makes sense for batch.Jobs (https://github.com/kubernetes-sigs/kueue/blob/b35ed012bee655264e6329b17c18fd756a08a4a7/pkg/features/kube_features.go#L95). We should probably default it to true and just keep it pending.

Reply:
Yeah sounds good :+1: